File: src/main/java/core/procedures/bit/BitFlip.java
package core.procedures.bit;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.Type;

public final class BitFlip extends AFn {

  public BitFlip() {
    super(new FnArgsBuilder().min(2).max(2).mandatory(new Class[] {Type.BitOp.class, Long.class}).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "bit-flip";
  }

  @Override
  public Long apply2(Object arg1, Object arg2) {
    long number = ((Number) arg1).longValue();
    return number ^ (1L << ((Number)arg2).longValue());
  }
}

File: src/main/java/core/procedures/bit/BitFlip.kt
package core.procedures.bit

import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.scm.Type

class BitFlip : AFn<Number?, Long>(name = "bit-flip", isPure = true, arity = Exactly(2),
                                   mandatoryArgsTypes = arrayOf(Type.BitOp::class.java, Long::class.javaObjectType)) {

    override operator fun invoke(arg1: Number?, arg2: Number?) = arg1!!.toLong() xor (1L shl arg2!!.toInt())
}
